### PR TITLE
thanos: Use wget for healthcheck

### PR DESCRIPTION
**Pull request by Reamer**
Use wget for healthcheck, because curl is not available in official thanos image (quay.io/thanos/thanos:v0.9.0)

### CODE REVIEW

**Comment by LiliC** [comment]
@Reamer hello, thanks for the PR! How did you encounter this problem, are you a customer of openshift or just tried it out as I don't see in the openshift dev team? :)

**Comment by Reamer** [comment]
Hi @LiliC,
I'm part of the [okd working group](https://github.com/openshift/community/blame/master/MEMBERS.md#L21) and checked some version of the okd 4.3 preview release. Done some experiments with updating components of okd and hit this issue.
`oc adm release new --from-image=....  --to-image=..  thanos=quay.io/thanos/thanos:v0.9.0`

**Comment by LiliC** [comment]
@Reamer Thanks for the clarification, that makes more sense now, as we didn't see any problems. Letting the tests run if no problems will merge, thanks for the PR! :)

**Comment by Reamer** [comment]
Hi @LiliC,
if anything with my code blocks the merge, please notify me.

**Comment by LiliC** [comment]
Sorry about that I forgot to click submit on the review right before vacation. :)

File: assets/thanos-querier/deployment.yaml [hunk 1]
```diff
@@ -33,14 +33,14 @@ spec:
...
-            - curl http://localhost:9090/-/healthy
+            - wget -O /dev/null http://localhost:9090/-/healthy
         name: thanos-querier
         readinessProbe:
           exec:
             command:
             - sh
             - -c
-            - curl http://localhost:9090/-/healthy
+            - wget -O /dev/null http://localhost:9090/-/healthy
```
[the remark this answers is not in the document]
Comment (s-urbaniak) [reply assets/thanos-querier/deployment.yaml]: I don't think we should not reference ipv4 addresses. I think wget first tries the ipv6 localhost address, fails, then retries the ipv4 address, which succeeds.

**Comment by s-urbaniak** [comment]
looking at the `e2e-aws-operator` test this simply seemed to have timed out. I don't see any errors in the logs and events.

**Comment by Reamer** [comment]
@LiliC I didn't test with IPv6 network stack. After adding IPv6 to my docker daemon. I tried starting `thanos query` with `--http-address=localhost:9090` to bind on both interfaces and to be equal with `wget` or `curl`-healthcheck. Unfortunately `thanos` didn't start on both interfaces
```
[root@575e34403f53 /]# ifconfig 
eth0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
...

lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
        inet6 ::1  prefixlen 128  scopeid 0x10<host>
        loop  txqueuelen 1000  (Local Loopback)
        RX packets 8  bytes 480 (480.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 8  bytes 480 (480.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

[root@575e34403f53 /]# netstat -tulpen
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       User       Inode      PID/Program name    
tcp        0      0 127.0.0.1:10901         0.0.0.0:*               LISTEN      99         203913     -                   
tcp        0      0 127.0.0.1:9090          0.0.0.0:*               LISTEN      99         203907     -   
```

I can reproduce your output, but the exit code of `wget` is still `0`. I think a change from `localhost` to `127.0.0.1` will makes no difference, except for the dns query. `localhost` should support all network stacks (ipv4 only, **ipv6 only** and dual stack)
```
[root@575e34403f53 /]# wget -O /dev/null http://localhost:9090/-/healthy
--2020-01-06 10:21:56--  http://localhost:9090/-/healthy
Resolving localhost (localhost)... ::1, 127.0.0.1
Connecting to localhost (localhost)|::1|:9090... failed: Connection refused.
Connecting to localhost (localhost)|127.0.0.1|:9090... connected.
HTTP request sent, awaiting response... 200 OK
Length: 23 [text/plain]
Saving to: '/dev/null'

100%[=========================================================================================================================================================================>] 23          --.-K/s   in 0s      

2020-01-06 10:21:56 (1.86 MB/s) - '/dev/null' saved [23/23]

[root@575e34403f53 /]# echo $?
0
```

**Comment by s-urbaniak** [comment]
Thanos won't listen on both ip addresses as golang itself won't bind to both ip addresses as per https://github.com/golang/go/issues/9334.

File: assets/thanos-querier/deployment.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (paulfantom) [reply assets/thanos-querier/deployment.yaml]: `localhost` is an exception and in this case it shouldn't matter if we use `127.0.0.1` or `localhost`, both should work the same.

FYI: we are using `127.0.0.1` in many places, for example in node-exporter daemonset.

**Comment by paulfantom** [comment]
I think if we want to have a proper healthcheck with `wget`, we should use something like:
```
wget --quiet --tries=1 --spider http://localhost:9090/-/healthy
```
Options passed to wget will reduce noise, try only once, and not download content.

**Comment by Reamer** [comment]
Confirm partial with @paulfantom. We should change to
```
wget --quiet --tries=1 --spider http://localhost:9090/-/healthy
```
I'll update my PR.
Disagree with 127.0.0.1 and localhost. We should not rely on a specific network stack. Yes I know 127.0.0.1 is used in many places.

**Comment by metalmatze** [comment]
You might want to reuse the work we already did in the Prometheus Operator.
That should work if either wget or curl is available.

https://github.com/coreos/prometheus-operator/blob/82d4fdcdc4067869c197910cb83d3ef10307d6e2/pkg/prometheus/statefulset.go#L619

**Comment by Reamer** [comment]
Hi @metalmatze 
thanks for sharing this information. I can adopt this solution. Why you are not using `--tries=1 --spider`? For shell command `command` you need a `bash`.

**Comment by paulfantom** [comment]
> I think for the shell command `command` you need at least a bash

Yes and no. `command` is a bash built-in and it is not included in "standard" posix shell. However I haven't found yet a system with posix shell in which `command` wouldn't be implemented (even busybox has it). In fact `command` is often recommended when writing portable shell scripts.

> Why don't you use --tries=1 --spider?

That's a good question.

**Comment by Reamer** [comment]
Changed PR to curl and wget. I hope that the multi line script doesn't make issues.

**Comment by LiliC** [comment]
@paulfantom do you mind reviewing this, thanks!

**Pull request merged**